### DISCUSSION TITLE
Is it worth it to have a feature request like a promise.all or future.wait like for AsyncValues?

### DISCUSSION THREAD

**Discussion by sandbox-apps** [rrousselGit/riverpod#715] [Q&A] [answered]
Guess that we have multiple asyncvalues per screen. In simplest scenario you would handle those things per entity. you place loading indicator per asyncvalue to handle their state. Im thinking that can we handle multiple AsyncValues combined into one?

**Comment by rrousselGit** [comment, marked as answer]
The difficulty is that most of the time, the combined AsyncValues return a different kind of object.

**Reply by iamarnas** [reply]
The easiest way to combine `AsyncValue` to one is by using `Tuple`.
#### Example:
```dart
AsyncValue<Tuple2<T, R>> combineAsync2<T, R>(
  AsyncValue<T> asyncOne,
  AsyncValue<R> asyncTwo,
) {
  if (asyncOne is AsyncError) {
    final error = asyncOne as AsyncError<T>;
    return AsyncError(error.error, error.stackTrace);
  } else if (asyncTwo is AsyncError) {
    final error = asyncTwo as AsyncError<R>;
    return AsyncError(error.error, error.stackTrace);
  } else if (asyncOne is AsyncLoading || asyncTwo is AsyncLoading) {
    return AsyncLoading();
  } else if (asyncOne is AsyncData && asyncTwo is AsyncData) {
    return AsyncData(Tuple2<T, R>(asyncOne.data.value, asyncTwo.data.value));
  } else {
    throw 'Unsupported case';
  }
}

AsyncValue<Tuple3<T, R, S>> combineAsync3<T, R, S>(
  AsyncValue<T> asyncOne,
  AsyncValue<R> asyncTwo,
  AsyncValue<S> asyncThree,
) {
//...
}
```

**Reply by rrousselGit** [reply]
That's the current solution indeed. But I don't want to include that in the package. It's still a flawed solution

**Reply by iamarnas** [reply]
I think you do not need include this to the package for few people, most of them create their own union classes with `Freezed`.

**Reply by chenyair** [reply]
> That's the current solution indeed. But I don't want to include that in the package. It's still a flawed solution

Do you think there will be an official solution soon? Because creating this method for each number of AsyncValues I wish to combine looks painful.

**Reply by davidmartos96** [reply]
> > That's the current solution indeed. But I don't want to include that in the package. It's still a flawed solution
> 
> Do you think there will be an official solution soon? Because creating this method for each number of AsyncValues I wish to combine looks painful.

There is an implementation in this thread that takes a list instead. https://github.com/rrousselGit/river_pod/issues/67#issuecomment-758670070

**Reply by rrousselGit** [reply]
There is a plan for a better solution. Although I am still in the process of looking for the correct solution.

**Reply by sandbox-apps** [reply]
I didn't know this will gain some atleast small discussion. Good to hear Im not just the only one in the community :D

**Reply by sandbox-apps** [reply]
This was came up because of the following
1. Some people do per entity solutions like for example using notifiers per entity.
2. Yes, you could create your own immutable class via freezed but others dont like to create and maintain immutable classes
That's why the easiest escape route for now is to combine asyncValues

**Reply by sandbox-apps** [reply]
I just wonder how can i refresh the specific error from provider

**Reply by iamarnas** [reply]
> There is an implementation in this thread that takes a list instead. https://github.com/rrousselGit/river_pod/issues/67#issuecomment-758670070

```dart
final history = watch(historyProvider);
final favorites = watch(favoritesProvider);

return merge([history, favorites]).when(
  data: (_) {
    print(history.unwrap());
    print(favorites.unwrap());
  } ,
  // ...
);
```
I'm not understand what a point to merge `history` and `favorites` when they are already initialized with `watch` and can uses in the tree below?
